### PR TITLE
add Autonomous TOTP

### PR DESCRIPTION
**Pull request by codenoid**

### CODE REVIEW

**Comment by rbsec** [comment]
Thanks for submitting this PR. We currently have an open PR for a new cheat sheet about MFA (#96 ), so this is perhaps content that would fit better in there. Once that new cheat sheet has been merged then we can just reference it from the Authentication cheat sheet (and a few other areas).

Do you think this section on TOTP has much that should be added to the TOTP section in the draft MFA cheat sheet?

Thanks

**Comment by ThunderSon** [comment]
I follow on with @rbsec on this. If your main point in your PR is about having it as a 2FA, I'd say that it should be added to the almost ready MFA CS. If you're talking about TOTP for authentication all time (very rare to be seen in the wild?), then it should stay in this CS.

**Comment by codenoid** [comment]
> If you're talking about TOTP for authentication all time (very rare to be seen in the wild?), then it should stay in this CS.

yes, it's about `TOTP for authentication all time`, as i know, one of unicorn-startup company uses this authentication method for their internal application

however, I still try to expand information about Autonomous TOTP

**Comment by mackowski** [comment]
Personally I am a big fan of small CS so you can also think about creating a CS about TOTP and link to it from Authentication and MFA. 
For now this is probably too small for a CS but I see a potential for this CS to grow over time.

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 1]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
 
 U2F augments password-based authentication using a hardware token (typically USB) that stores cryptographic authentication keys and uses them for signing. The user can use the same token as a second factor for multiple applications. U2F works with web applications. It provides **protection against phishing** by using the URL of the website to lookup the stored authentication key.
 
+## Autonomous TOTP
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: This isn't really a term I've seen before - maybe just `TOTP` would be better?

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 2]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
...
+
+The Time-based One-Time Password algorithm (TOTP) is an extension of the HMAC-based One-time Password algorithm (HOTP) generating a one-time password by instead taking uniqueness from the current time. It has been adopted as Internet Engineering Task Force[1] standard RFC 6238, is the cornerstone of Initiative For Open Authentication (OATH), and is used in a number of [two-factor authentication](https://en.wikipedia.org/wiki/Two-factor_authentication) systems.
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: This is all correct, but seems to be a bit to dry and technical for the cheat sheet. It's good to have a link where readers can learn more about the technical details of TOTP, but this doesn't seem like stuff that everyone needs to know.

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 3]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
...
+
+A server is needed to generating token for TOTP Value, the client must have same logic with their authentication server, so *authorized client* should be always send a valid TOTP Value
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: This sentence doesn't really seem to add anything

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 4]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
...
+
+Autonomous TOTP must meet several requirements, namely:
+
+1. token are created by machine dynamically (autonomously), otherwise you need to store primary token safely
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: I'm not really sure what this is saying?

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 5]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
...
+2. connection with low latency is needed
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: Is this true? AFAIK TOTP tokens are usually valid for 30s, and you can normally use one token on either side, so there's a reasonable amount of leeway. Perhaps rephrase as something like `The time of the two systems needs to be relatively closely in sync.`?

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 6]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
...
+3. secure connection is forced to transfer the token
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: Maybe `The token should be submitted over a secure (HTTPS) connection.`?

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 7]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
 
 U2F augments password-based authentication using a hardware token (typically USB) that stores cryptographic authentication keys and uses them for signing. The user can use the same token as a second factor for multiple applications. U2F works with web applications. It provides **protection against phishing** by using the URL of the website to lookup the stored authentication key.
 
+## Autonomous TOTP
+
+The Time-based One-Time Password algorithm (TOTP) is an extension of the HMAC-based One-time Password algorithm (HOTP) generating a one-time password by instead taking uniqueness from the current time. It has been adopted as Internet Engineering Task Force[1] standard RFC 6238, is the cornerstone of Initiative For Open Authentication (OATH), and is used in a number of [two-factor authentication](https://en.wikipedia.org/wiki/Two-factor_authentication) systems.
+
+A server is needed to generating token for TOTP Value, the client must have same logic with their authentication server, so *authorized client* should be always send a valid TOTP Value
+
+Autonomous TOTP must meet several requirements, namely:
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: Are these hard requirements to implement TOTP, or best practices for how it should be done?

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 8]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
...
+
+1. token are created by machine dynamically (autonomously), otherwise you need to store primary token safely
+2. connection with low latency is needed
+3. secure connection is forced to transfer the token
+4. do not ever only use token that are transferred to generate a valid TOTP Value, developer must be generate a new token based on transferred token
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: I don't understand this sentence

File: cheatsheets/Authentication_Cheat_Sheet.md [hunk 9]
```diff
@@ -270,6 +270,21 @@ UAF takes advantage of existing security technologies present on devices for aut
...
+
+Autonomous TOTP makes request replication is more harder for hacker, even though the hacker gets a token to create a TOTP Value, he may not be able to generate a valid TOTP Value.
```
Comment (rbsec) [inline cheatsheets/Authentication_Cheat_Sheet.md]: Again, I'm afraid I'm not really sure what this is saying.

**Comment by codenoid** [comment]
i'm too lazy for this, hope someone makes a better version of this cs, and sorry for your time

**Pull request closed without merging**